tests/loggen/file_reader.c

```diff
@@ -0,0 +1,239 @@
+  { "skip-tokens", 0, 0, G_OPTION_ARG_INT, &skip_tokens, "Skip the given number of tokens (delimined by a space) at the beginning of each line (default value: 3)", "<number>" },
+  { "dont-parse", 'd', 0, G_OPTION_ARG_NONE, &dont_parse, "Don't parse the lines coming from the readed files. Loggen will send the whole lines as it is in the readed file", NULL },
+  { NULL }
+};
+
+FILE **source; // array for file handlers
+
+GOptionEntry *
+get_file_reader_options(void)
+{
+  return loggen_file_reader_options;
+}
+
+int init_file_reader(int nr_threads)
+{
+  if (!read_file_name)
+    {
+      DEBUG("file reader not activated\n");
+      return 0;
+    }
+
+  source = malloc(nr_threads * sizeof(FILE *));
+
+  if (read_file_name[0] == '-' && read_file_name[1] == '\0')
+    {
+      for (int i=0; i<nr_threads; i++)
+        source[i] = stdin;
+    }
+  else
+    {
+      for (int i=0; i<nr_threads; i++)
+        {
+          source[i] = fopen(read_file_name, "r");
+          if (!source[i])
+            {
+              ERROR("unable to open file (%s)\n",read_file_name);
+              return -1;
+            }
+        }
+
+      DEBUG("file (%s) opened for reading\n",read_file_name);
+    }
+  return 1;
+}
+
+void close_file_reader(int nr_threads)
+{
+  if (!source)
+    return;
+
+  for (int i=0; i<nr_threads; i++)
+    {
+      if (source[i])
+        fclose(source[i]);
+      source[i] = NULL;
+    }
+  free(source);
+}
+
+static int
+parse_line(const char *line, char *host, char *program, char *pid, char **msg)
+{
+  const char *pos0;
+  const char *pos = line;
+  const char *end;
+  int space = skip_tokens;
+  int pid_len;
+
+  /* Find token */
+  while (space)
+    {
+      pos = strchr(pos, ' ');
+
+      if (!pos)
+        return -1;
+      pos++;
+
+      space --;
+    }
+
+  pos = strchr(pos, ':');
+  if (!pos)
+    return -1;
+
+  /* pid */
+  pos0 = pos;
+  if (*(--pos) == ']')
+    {
+      end = pos - 1;
+      while (*(--pos) != '[')
+        ;
+
+      pid_len = end -
+                pos; /* 'end' points to the last character of the pid string (not off by one), *pos = '[' -> pid length = end - pos*/
+      memcpy(pid, pos + 1, pid_len);
+      pid[pid_len] = '\0';
+    }
+  else
+    {
+      pid[0] = '\0';
+      ++pos; /* the 'pos' has been decreased in the condition (']'), reset it to the original position */
+    }
+
+  /* Program */
+  end = pos;
+  while (*(--pos) != ' ')
+    ;
+
+  memcpy(program, pos + 1, end - pos - 1);
+  program[end-pos-1] = '\0';
+
+  /* Host */
+  end = pos;
+  while (*(--pos) != ' ')
+    ;
+
+  memcpy(host, pos + 1, end - pos - 1);
+  host[end-pos-1] = '\0';
+
+  *msg = ((char *)pos0) + 2;
+
+  return 1;
+}
+
+int
+read_next_message_from_file(char *buf, int buflen, int syslog_proto, int thread_index)
+{
+  static int lineno = 0;
+  int linelen;
+
+  char host[128], program[128], pid[16];
+  char *msg = NULL;
+
+  if (!source[thread_index])
+    return 0;
+
+  while (1)
```

Comment:
This is a pretty old legacy code part. I just reused it. I know this is not an excuse :-)
Anyhow I think I really need this while because we may need to read file in a loop (--loop-reading option)